opensanctions/crawlers/us_congress.py

```diff
@@ -0,0 +1,122 @@
+@cache
+def to_date(days: int) -> str:
+    dt = settings.RUN_TIME - timedelta(days=days)
+    return dt.isoformat()[:10]
+
+
+def make_occupancy(
+    context: Context,
+    person,
+    position,
+    start_date,
+    end_date: Optional[str],
+    no_end_implies_current: bool,
+) -> Optional[Entity]:
+    """Helper for making occupancies if they meet our criteria for PEP position occupancy"""
+    # TODO: Change dates to be dates to generalise
+    # TODO: The general version will need more of the logic from wd_peps than just this clause
+    if end_date is None or end_date > to_date(AFTER_OFFICE):
+        occupancy = context.make("Occupancy")
+        parts = [person.id, position.id, start_date, end_date]
+        occupancy.id = context.make_id(*parts)
+        occupancy.add("holder", person)
+        occupancy.add("post", position)
+        occupancy.add("startDate", start_date)
+        occupancy.add("endDate", end_date)
+        if no_end_implies_current and not end_date:
+            status = "Current"
+        else:
+            status = "Ended"
+        occupancy.add("status", status)
+        return occupancy
+
+
+def crawl_positions(context, member, entity):
+    terms: List[dict] = member.pop("terms")
+    entities = []
+    topics = set()
+    for term in terms:
+        res = context.lookup("position", term["chamber"])
+
+        position = make_position(context, res.name, country="us")
+        occupancy = make_occupancy(
+            context,
+            entity,
+            position,
+            str(term.pop("startYear")),
+            str(term.pop("endYear", None)),
+            True,
+        )
+        if occupancy:
+            entities.append(position)
+            entities.append(occupancy)
+            topics.update(res.topics)
+    return entities, topics
+
+
+def crawl_member(context: Context, bioguide_id: str):
+    url = urljoin(context.source.data.url, bioguide_id)
+    headers = {"x-api-key": API_KEY}
+    member = context.fetch_json(url, headers=headers, cache_days=CACHE_DAYS)["member"]
+
+    person = context.make("Person")
+    person.id = context.make_id(bioguide_id)
+    person.add(
+        "name", member.pop("directOrderName")
+    )  # TODO: use name parts from full member endpoint
+    person.add("country", "us")
+
+    entities, topics = crawl_positions(context, member, person)
+    if entities:
+        person.add("topics", topics)
+        context.emit(person, target=True)
+        for entity in entities:
+            context.emit(entity)
+
+
+def fetch(context: Context, offset):
+    query = {"limit": LIMIT, "offset": offset}
+    url = f"{ context.source.data.url }?{ urlencode(query) }"
+    headers = {"x-api-key": API_KEY}
+    path = context.fetch_resource(f"members-{offset}.json", url, headers=headers)
+    context.export_resource(path, title=context.SOURCE_TITLE + f"offset {offset}")
+    with open(path, "r") as fh:
+        return json.load(fh)["members"], offset + LIMIT
```

Comment:
is this useful/worth it when we're actually just using it for getting IDs then making /member/id requests for the actual data? when do we bother exporting the resource and when don't we?

Reply:
It's a bit fuzzy but the reason for keeping the exported resources is that we can reasonably write a script later to re-generate the data from it (or run a modified version of the crawler on historic data). For example, for the US sanctions list, we could relatively easy go and parse each version of the data released in the last 2 years to get fully accurate time stamps on entities. For a more complex crawler like this, it's really unlikely we'll ever be able to do time travel. We also don't want to show the 600+ resources in the dataset page on OpenSanctions. So I'd skip it.